src/addrman.h

```diff
@@ -268,6 +275,46 @@ class CAddrMan
     void SetServices_(const CService &addr, ServiceFlags nServices) EXCLUSIVE_LOCKS_REQUIRED(cs);
 
 public:
+    // Compressed IP->ASN mapping, loaded from a file when a node starts.
+    // Should be always empty if no file was provided.
+    // This mapping is then used for bucketing nodes in Addrman.
+    //
+    // If asmap is provided, nodes will be bucketed by
+    // AS they belong to, in order to make impossible for a node
+    // to connect to several nodes hosted in a single AS.
+    // This is done in response to Erebus attack, but also to generally
+    // diversify the connections every node creates,
+    // especially useful when a large fraction of nodes
+    // operate under a couple of cloud providers.
+    //
+    // If a new asmap was provided, the exsiting records
+    // would be re-bucketed accordingly.
+    std::vector<bool> m_asmap;
+
+    static std::vector<bool> DecodeAsmap(fs::path path)
+    {
+        std::vector<bool> bits;
+        FILE *filestr = fsbridge::fopen(path, "rb");
+        CAutoFile file(filestr, SER_DISK, CLIENT_VERSION);
+        if (file.IsNull()) {
+            LogPrintf("Failed to open asmap file from disk.\n");
+            return bits;
+        }
+        fseek(filestr, 0, SEEK_END);
+        int length = ftell(filestr);
+        LogPrintf("Opened asmap file (%d bytes) from disk.\n", length);
```

Comment:
Suggest to provide feedback on which filename was opened:
```diff
-        LogPrintf("Opened asmap file (%d bytes) from disk.\n", length);
+        LogPrintf("Opened asmap file %s (%d bytes) from disk.\n", path, length);
```
Tested examples of output:
```
Opened asmap file "/home/jon/.bitcoin/ip_asn.map" (932999 bytes) from disk.
Opened asmap file "/home/jon/.bitcoin/../projects/bitcoin/asmap/demo.map" (932999 bytes) from disk.
```